Avoid application deletion race conditions

thepetk opened the pull request:
### What does this PR do?:
This PR tries to implement the same logic followed to the `component_controller` (introduced here: https://github.com/redhat-appstudio/application-service/pull/451) on the `application_controller` too. This way we avoid race conditions upon `application` & `component` deletion.

More detailed, we are ensuring that before the `application` deletion (after we have confirmed the presence of a finalizer), that the application has no components under deletion. If it does we are requeueing the deletion.

### Which issue(s)/story(ies) does this PR fixes:
https://issues.redhat.com/browse/DEVHAS-627

### PR acceptance criteria:
<!--Testing and documentation do not need to be complete in order for this PR to be approved. We just need to ensure tracking issues are opened.

> - Open new test/doc issues
> - Check each criteria if:
>  - There is a separate tracking issue. Add the issue link under the criteria
>  **or**
>  - test/doc updates are made as part of this PR
> -  If unchecked, explain why it's not needed
-->

- [ ] Unit/Functional tests

  <!-- _These are run as part of the PR workflow, ensure they are updated_ -->

- [ ] Documentation 

   <!-- _This includes product docs and READMEs._ -->

- [ ] Client Impact

  <!-- _Do we have anything that can break our clients?  If so, open a notifying issue_ -->


### How to test changes / Special notes to the reviewer:

thepetk commented:
/retest

maysunfaisal commented on controllers/application_controller.go:
```diff
@@ -114,6 +114,16 @@ func (r *ApplicationReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	} else {
 		if containsString(application.GetFinalizers(), appFinalizerName) {
+			var components appstudiov1alpha1.ComponentList
+			err = r.List(context.Background(), &components)
```
```suggestion
			err = r.List(ctx &components)
```

maysunfaisal commented on controllers/application_controller.go:
```diff
@@ -114,6 +114,16 @@ func (r *ApplicationReconciler) Reconcile(ctx context.Context, req ctrl.Request)
+			var components appstudiov1alpha1.ComponentList
+			err = r.List(context.Background(), &components)
+			if err != nil {
+				log.Error(err, "failed to check if any application component is under deletion")
```
```suggestion
				log.Error(err, "failed to check if any Component is under deletion, proceeding with Application delete")
```

maysunfaisal commented on controllers/application_controller.go:
```diff
@@ -114,6 +114,16 @@ func (r *ApplicationReconciler) Reconcile(ctx context.Context, req ctrl.Request)
+			}
+			if err := util.VerifyNoApplicationComponentUnderDeletion(components, application.Name); err != nil {
+				// requeue in case any of the application components are under deletion
+				log.Error(err, "Error an application component is under deletion. Deletion requeued")
```
```suggestion
				log.Error(err, "Component is under deletion, Application deletion requeued")
```

maysunfaisal reviewed (commented):
small changes

maysunfaisal reviewed (approved):
/lgtm

maysunfaisal commented:
/retest

johnmcollier commented:
> More detailed, we are ensuring that before the application deletion (after we have confirmed the presence of a finalizer), that the application has no components under deletion. If it does we are requeueing the deletion.

I'm wondering if this is needed? Since we fixed the issue where the Component deletion could get stuck if the GitOps repository has been deleted, I don't think there's a potential race condition that can occur here?

johnmcollier reviewed (commented):
> More detailed, we are ensuring that before the application deletion (after we have confirmed the presence of a finalizer), that the application has no components under deletion. If it does we are requeueing the deletion.

I'm wondering if this is needed? Since we fixed the issue where the Component deletion could get stuck if the GitOps repository has been deleted, I don't think there's a potential race condition that can occur here? If there's Components that are under deletion when we delete the Application, that just means the GitOps repository will get deleted before the Component finalizer runs (and since https://github.com/redhat-appstudio/application-service/pull/451 was merged, that shouldn't be an issue).

thepetk commented:
> > More detailed, we are ensuring that before the application deletion (after we have confirmed the presence of a finalizer), that the application has no components under deletion. If it does we are requeueing the deletion.
> 
> I'm wondering if this is needed? Since we fixed the issue where the Component deletion could get stuck if the GitOps repository has been deleted, I don't think there's a potential race condition that can occur here? If there's Components that are under deletion when we delete the Application, that just means the GitOps repository will get deleted before the Component finalizer runs (and since #451 was merged, that shouldn't be an issue).
> 
> We remove the finalizer no matter what [here](https://github.com/thepetk/application-service/blob/a08a1287423736befe5cebdedb70bf100616f4db/controllers/component_controller.go#L210-L222), so even if the Component clean up "fails" due to a missing GitOps repository, the deletion of either resources won't be blocked



> > More detailed, we are ensuring that before the application deletion (after we have confirmed the presence of a finalizer), that the application has no components under deletion. If it does we are requeueing the deletion.
> 
> I'm wondering if this is needed? Since we fixed the issue where the Component deletion could get stuck if the GitOps repository has been deleted, I don't think there's a potential race condition that can occur here? If there's Components that are under deletion when we delete the Application, that just means the GitOps repository will get deleted before the Component finalizer runs (and since #451 was merged, that shouldn't be an issue).
> 
> We remove the finalizer no matter what [here](https://github.com/thepetk/application-service/blob/a08a1287423736befe5cebdedb70bf100616f4db/controllers/component_controller.go#L210-L222), so even if the Component clean up "fails" due to a missing GitOps repository, the deletion of either resources won't be blocked

FYI the PR is basically the extra step discussed here: https://github.com/redhat-appstudio/application-service/pull/451#discussion_r1529559854

The main reason is that we want to fix the component deletion metrics. I've tried to revisit the investigation on the component deletion errors after your comment. My notes:

* The truth is In the latest error logs I cannot see any `Unable to update GitOps repository for component..` which comes from an unsuccessful deletion.

* As we have already noticed the main log is: `Operation cannot be fulfilled on...the object has been modified;`. I think this error comes from [line#213](https://github.com/redhat-appstudio/application-service/blob/main/controllers/component_controller.go#L213) where we are trying to update the `to-be-deleted` component.

* I can see that in [line#194](https://github.com/redhat-appstudio/application-service/blob/main/controllers/component_controller.go#L194) we are increasing the total deletion attempts. But we only increase the component deletion failed/succeeded for specific cases (see [line#205](https://github.com/redhat-appstudio/application-service/blob/main/controllers/component_controller.go#L205) & [line#219](https://github.com/redhat-appstudio/application-service/blob/main/controllers/component_controller.go#L219)).

* Having in mind the above note I think the reason of the component deletion not successful alert might be coming because our total attempts are much bigger than the sum of failed + succeeded (e.g we increase the total attempts but we don't equally increase the failed/successful). As a result the succeeded attempts should be only a small pertecantge of the total deletion attempts.

From the above notes I'd say we could:

* Increase total attempts only for cases we decide that the attempt is failed/succeded
OR
* We add succeded attempts for lines [214](https://github.com/redhat-appstudio/application-service/blob/main/controllers/component_controller.go#L214) & [216](https://github.com/redhat-appstudio/application-service/blob/main/controllers/component_controller.go#L216)

wdyt? cc @maysunfaisal @johnmcollier

thepetk commented:
> > > More detailed, we are ensuring that before the application deletion (after we have confirmed the presence of a finalizer), that the application has no components under deletion. If it does we are requeueing the deletion.
> > 
> > 
> > I'm wondering if this is needed? Since we fixed the issue where the Component deletion could get stuck if the GitOps repository has been deleted, I don't think there's a potential race condition that can occur here? If there's Components that are under deletion when we delete the Application, that just means the GitOps repository will get deleted before the Component finalizer runs (and since #451 was merged, that shouldn't be an issue).
> > We remove the finalizer no matter what [here](https://github.com/thepetk/application-service/blob/a08a1287423736befe5cebdedb70bf100616f4db/controllers/component_controller.go#L210-L222), so even if the Component clean up "fails" due to a missing GitOps repository, the deletion of either resources won't be blocked
> 
> > > More detailed, we are ensuring that before the application deletion (after we have confirmed the presence of a finalizer), that the application has no components under deletion. If it does we are requeueing the deletion.
> > 
> > 
> > I'm wondering if this is needed? Since we fixed the issue where the Component deletion could get stuck if the GitOps repository has been deleted, I don't think there's a potential race condition that can occur here? If there's Components that are under deletion when we delete the Application, that just means the GitOps repository will get deleted before the Component finalizer runs (and since #451 was merged, that shouldn't be an issue).
> > We remove the finalizer no matter what [here](https://github.com/thepetk/application-service/blob/a08a1287423736befe5cebdedb70bf100616f4db/controllers/component_controller.go#L210-L222), so even if the Component clean up "fails" due to a missing GitOps repository, the deletion of either resources won't be blocked
> 
> FYI the PR is basically the extra step discussed here: [#451 (comment)](https://github.com/redhat-appstudio/application-service/pull/451#discussion_r1529559854)
> 
> The main reason is that we want to fix the component deletion metrics. I've tried to revisit the investigation on the component deletion errors after your comment. My notes:
> 
> * The truth is In the latest error logs I cannot see any `Unable to update GitOps repository for component..` which comes from an unsuccessful deletion.
> * As we have already noticed the main log is: `Operation cannot be fulfilled on...the object has been modified;`. I think this error comes from [line#213](https://github.com/redhat-appstudio/application-service/blob/main/controllers/component_controller.go#L213) where we are trying to update the `to-be-deleted` component.
> * I can see that in [line#194](https://github.com/redhat-appstudio/application-service/blob/main/controllers/component_controller.go#L194) we are increasing the total deletion attempts. But we only increase the component deletion failed/succeeded for specific cases (see [line#205](https://github.com/redhat-appstudio/application-service/blob/main/controllers/component_controller.go#L205) & [line#219](https://github.com/redhat-appstudio/application-service/blob/main/controllers/component_controller.go#L219)).
> * Having in mind the above note I think the reason of the component deletion not successful alert might be coming because our total attempts are much bigger than the sum of failed + succeeded (e.g we increase the total attempts but we don't equally increase the failed/successful). As a result the succeeded attempts should be only a small pertecantge of the total deletion attempts.
> 
> From the above notes I'd say we could:
> 
> * Increase total attempts only for cases we decide that the attempt is failed/succeded
>   OR
> * We add succeded attempts for lines [214](https://github.com/redhat-appstudio/application-service/blob/main/controllers/component_controller.go#L214) & [216](https://github.com/redhat-appstudio/application-service/blob/main/controllers/component_controller.go#L216)
> 
> wdyt? cc @maysunfaisal @johnmcollier

as discussed I'm closing for now the PR. If in the future this solution is needed we can re-open.
